### PR TITLE
Introduce GiteaLocaleNumber custom element to handle number localization on pages.

### PR DESCRIPTION
**Pull request by wxiaoguang**
Follow #21429 & #22861

Use `<gitea-locale-number>` instead of backend `PrettyNumber`. All old `PrettyNumber` related functions are removed. A lot of code could be simplified. 

And some functions haven't been used for long time (dead code), so they are also removed by the way (eg: `SplitStringAtRuneN`, `Dedent`)

This PR only tries to improve the `PrettyNumber` rendering problem, it doesn't touch the "plural" problem.

Screenshot:

image

image

### CODE REVIEW

**Comment by silverwind** [comment]
> the "plural" problem

Maybe via https://2ality.com/2019/12/intl-pluralrules.html, but as the post mentions, ideally it should be translation files as the method is likely not perfect.

**Comment by wxiaoguang** [comment]
> > the "plural" problem
> 
> Maybe https://2ality.com/2019/12/intl-pluralrules.html is a start.

To resolve the  "plural" problem clearly, first we must have a clear translation definition & syntax.

Although there is a PR #19916 trying to resolve the plural translation, but it seems much too complicated, I am afraid that it's not friendly to translators.

File: web_src/js/webcomponents/GiteaLocaleNumber.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    if (number) {
+      this.attachShadow({mode: 'open'});
+      this.shadowRoot.textContent = new Intl.NumberFormat().format(Number(number));
+    }
+    let numberInTooltip = this.getAttribute('data-number-in-tooltip');
+    if (numberInTooltip) {
+      numberInTooltip = JSON.parse(numberInTooltip);
+      let msg = numberInTooltip.message;
+      msg = msg.replace(/%[ds]/, new Intl.NumberFormat().format(Number(numberInTooltip.number)));
+      this.setAttribute('data-tooltip-content', msg);
```
Comment (silverwind) [inline web_src/js/webcomponents/GiteaLocaleNumber.js]:
```suggestion
      const {message} = JSON.parse(numberInTooltip);
      this.setAttribute('data-tooltip-content', message
        .replace(/%[ds]/, new Intl.NumberFormat().format(Number(numberInTooltip.number)))
      );
```

no mutations is better :)

Comment (wxiaoguang) [reply web_src/js/webcomponents/GiteaLocaleNumber.js]: Good point (ps: the suggested change seems not quite right 😁 ). There is a `numberInTooltip.number` access.

I will propose new changes.

----

Done in 80844f8d8dffd844fe140c251b0c3b979a3e81d3

**Review by lunny — APPROVED** [review]

File: web_src/js/webcomponents/GiteaLocaleNumber.js (hunk 1 above)
Comment (silverwind) [reply web_src/js/webcomponents/GiteaLocaleNumber.js]: Ah yes, overlooked that below, looks good.

**Review by silverwind — APPROVED** [review]

**Comment by jolheiser** [comment]
🎺 🤖

**Comment by zeripath** [comment]
> > > the "plural" problem
> > 
> > 
> > Maybe https://2ality.com/2019/12/intl-pluralrules.html is a start.
> 
> To resolve the "plural" problem clearly, first we must have a clear translation definition & syntax.
> 
> Although there is a PR #19916 trying to resolve the plural translation, but it seems much too complicated, I am afraid that it's not friendly to translators.

Inherently plural rules are complex.

I have tried multiple times to suggest options for making things less complicated but the PR has simply languished without review or interest.

**Comment by silverwind** [comment]
Maybe I don't understand the problem fully. For english-only I usually get away with https://github.com/plurals/pluralize to transform a noun into singular or plural. But english is simple, other languages are more complex I assume.

**Comment by wxiaoguang** [comment]
I have taken a try for the "plural" problem recently, according to my knowledge #23863, my options are (at the moment):

1. we do not need to re-invent the wheel (parsing the CLDR rules) , the official `x/text` package works well.
2. we need to use a very simple and clear syntax for strings, otherwise the translators won't write correctly, they would break the string in the end.
    * I have found many broken translations, and I am trying to fix them
3. we need to improve our locale package first, fix the parsing & syncing problem with Crowdin

**Comment by zeripath** [comment]
When I wrote the referenced PR we very much did need to reinvent the parsing of cldr rules because `x/text` did not do it. Perhaps it has improved in the extremely long time it has been since I wrote that pr.

**Pull request merged**